_posts/2019-07-27-GettingStartedWithProtobufGrpc.md

```diff
@@ -0,0 +1,306 @@
+- **gRPC Protcol** : an HTTP/2 based protocol to enable client-server communication.
+
+![gRPC client-server communication](/images/2019-07-27-GettingStartedWithProtobufGrpc/gRPC.svg)
+[Source](https://grpc.io/)
+
+Using these two components, gRPC enables applications to communicate over HTTP/2 tcp connection while sending data in binary encoded format. This is a great way to create microservices and systems that scale very well horizontally. Moreover, the framework is highly performant, avoids the need to write boilerplate code and works seamlessly on the cloud without any hacks. One of the most promising features of gRPC is that it is programing language agnostic. You can start playing around with gRPC with any programming language of your choice.
+
+From gRPC website:
+
+> gRPC is a modern, open source remote procedure call (RPC) framework that can run anywhere. It enables client and server applications to communicate transparently and makes it easier to build connected systems.
+
+
+## Content:
+
+1. What is gRPC and why would we use it? 
+1.1. Features. 
+1.2. Supported programming languages. 
+1.3. Supported data formats.
+1.4. Typical applications that can be built with gRPC.
+
+2. Fitting pieces together. [General workflow for a simple project]
+2.1. Defining payload and service definition.
+2.2. Generate gRPC code from protobuf definition.
+2.3. Implement gRPC server.
+2.4. Implement gRPC client.
+2.5. Run the server and call it using a client.
+
+3. Where to go next.
+4. Sources and References.
+
+## 1. What is gRPC and why would we use it?
+
+If you have worked with REST APIs, what you are generally used to is:
+```
+REST API = HTTP1.1 + JSON + REST
+```
+where `HTTP1.1` is the transport protocol, `JSON` is message format, and `REST` is the architectural style (resourceful endpoint).
+
+gRPC handles the same scenario as:
+```
+gRPC API = HTTP/2 + Protobuf + RPC
+```
+Here the transport protocol is based on `HTTP/2`, the data format is defined using `Protobuf` and architectural style is `RPC`.
+
+Few advantages of this approach are:
+- 1. **Transport level optimization**: `HTTP/2` is more performant that standard `HTTP1.1`. One of the major advantages is that HTTP/2 creates a single connection between client and server and both parties can exchange data over a single connection, whereas with HTTP1.1 each data exchange needs a separate connection. Check the link at end of this section for more detailed explanation.
+- 2. **Efficient serialization**: `Protobuf` is a language-neutral way for serializing structured data. They are also very efficient and packed, while JSON, on the other hand, is textual inherently. If you have worked with JSON and XML, Protobuf lies in the same space with an added advantage that protobuf compiler can translate your data definitions to the language of your choice. Therefore, you create your structure once and use it across all supported languages.
```

Comment:
What do you mean by "If you have worked with JSON and XML, Protobuf lies on the same space"? If the previous statement says that "while JSON, on the other hand ..."
I'd say they are in different spaces, Protobuf would be in the same space as Avro or Thrift: they use binary formats.

Reply:
What I meant was describing data in a specific format. But I see that the statement is confusing. Will rephrase it.